File: CivitasRuby/vista_textual.rb
#encoding:utf-8
require_relative 'operaciones_juego'
require_relative 'civitas_juego'
require_relative 'jugador'
require_relative 'diario'
require 'io/console'

module Civitas

  class Vista_textual

    def mostrar_estado(estado)
      puts estado
    end
    
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)
            numero = cadena.to_i
            ok = true
          else
            raise IOError
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end

    def menu(titulo,lista)
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opcion: ",
                          tab+"Valor erroneo")
      return opcion
    end

    def salir_carcel
      opcion = menu("Elige la forma para intentar salir de la carcel", ["Pagando", "Tirando el dado"])
      Lista_SalidasCarcel[opcion]
    end
    
    def comprar
      respuestas = []
      for i in 0..Lista_Respuestas.size-1
        respuestas.push(Lista_Respuestas[i].to_s)
      end
      
      opcion = menu("¿Desea comprar la propiedad?", respuestas)
      Lista_Respuestas[opcion]
    end

    def gestionar
      gestiones = []
      for i in 0..Lista_GestionesInmobiliarias.size-1
        gestiones.push(Lista_GestionesInmobiliarias[i].to_s)
      end
      @iGestion = menu("Introduzca que gestion desea realizar", gestiones)
		
      if @iGestion != 5
        str_propiedades = []
        propiedades = @juegoModel.get_jugador_actual.propiedades
        
        for i in 0..propiedades.size-1
          cadena = propiedades[i].nombre + ". Precio: " + propiedades[i].precioCompra.to_s
          str_propiedades.push(cadena)
        end

        @iPropiedad = menu("Seleccione que propiedad gestionar", str_propiedades)
      end
    end

    def getGestion
      @iGestion
    end

    def getPropiedad
      @iPropiedad
    end

    def mostrarSiguienteOperacion(operacion)
      puts("La siguiente operacion a realizar es: " + operacion.to_s)
    end

    def mostrarEventos
      while Diario.instance.eventos_pendientes
        puts(Diario.instance.leer_evento)
      end
    end

    def setCivitasJuego(civitas)
         @juegoModel=civitas
         self.actualizarVista
    end

    def actualizarVista
      jugador = @juegoModel.get_jugador_actual
      num_propiedades = jugador.propiedades.size
      casilla_actual = @juegoModel.get_casilla_actual
		
      info = "Jugador actual: " + jugador.nombre.to_s
			info += "\nCasilla: " + casilla_actual.nombre
			info += "\nNum. propiedades: " + num_propiedades.to_s + "\n"
		
      for i in 0..num_propiedades-1
        propiedades = jugador.propiedades
        info += i.to_s + ") " + propiedades[i].to_s + "\n"
      end
      info += "\n"
    end
  end
end

File: CivitasJava/juegoTexto/VistaTextual.java
package juegoTexto;

import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.SalidasCarcel;
import civitas.Respuestas;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.Casilla;
import civitas.Jugador;
import civitas.TituloPropiedad;

class VistaTextual{
	CivitasJuego juegoModel; 
	int iGestion=-1;
	int iPropiedad=-1;
	private static String separador = "=====================";
	private Scanner in;
	
	VistaTextual(){
            in = new Scanner(System.in);
	}
	
	void mostrarEstado(String estado){
            System.out.println(estado);
	}
	
	void pausa() {
            System.out.print ("Pulsa una tecla");
            in.nextLine();
	}

	int leeEntero (int max, String msg1, String msg2){
            boolean ok;
            String cadena;
            int numero = -1;
            do{
                System.out.print (msg1);
                cadena = in.nextLine();
                try{
                    numero = Integer.parseInt(cadena);
                    ok = true;
                }catch(NumberFormatException e){
                    System.out.println(msg2);
                    ok = false;  
                }
                if (ok && (numero < 0 || numero >= max)){
                    System.out.println(msg2);
                    ok = false;
                }
            }while (!ok);

            return numero;
	}

	int menu (String titulo, ArrayList<String> lista){
            String tab = "  ";
            int opcion;
            System.out.println(titulo);
            for (int i = 0; i < lista.size(); i++) {
                System.out.println(tab+i+"-"+lista.get(i));
            }

            opcion = leeEntero(lista.size(),
                    "\n"+tab+"Elige una opción: ",
                    tab+"Valor erróneo");
		
            return opcion;
	}

	SalidasCarcel salirCarcel() {
            int opcion = menu ("Elige la forma para intentar salir de la carcel", new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
            return (SalidasCarcel.values()[opcion]);
	}

	Respuestas comprar(){
            ArrayList<String> opciones = new ArrayList<>(Arrays.asList("SI", "NO"));
            int opcion = menu("¿Desea comprar la propiedad?", opciones);

            return (Respuestas.values()[opcion]);
	}

	void gestionar(){
            ArrayList<String> opciones = new ArrayList<>(Arrays.asList("VENDER", "HIPOTECAR", "CANCELAR_HIPOTECA", "CONSTRUIR_CASA", "CONSTRUIR_HOTEL",	"TERMINAR"));
            iGestion = menu("Introduzca que gestion desea realizar", opciones);

            if (iGestion != 5){
                ArrayList<String> strPropiedades = new ArrayList<>();
                ArrayList<TituloPropiedad> propiedades = new ArrayList<>(juegoModel.getJugadorActual().getPropiedades());

                for (int i = 0; i < propiedades.size(); i++)
                    strPropiedades.add(propiedades.get(i).getNombre() );

                iPropiedad = menu("Seleccione que propiedad gestionar", strPropiedades);
            }
	}
  
	public int getGestion(){
            return iGestion;
	}
	
	public int getPropiedad(){
            return iPropiedad;
	}
    
	void mostrarSiguienteOperacion(OperacionesJuego operacion){
            System.out.println("La siguiente operación a realizar es: " + operacion.toString());
	}

	void mostrarEventos(){
            while (Diario.getInstance().eventosPendientes()){
                System.out.println(Diario.getInstance().leerEvento());
            }
	}
  
	public void setCivitasJuego(CivitasJuego civitas){ 
            juegoModel = civitas;
            this.actualizarVista();
	}
  
	void actualizarVista(){
            Jugador jugador = juegoModel.getJugadorActual();
            int numPropiedades = jugador.getPropiedades().size();
            Casilla casillaActual = juegoModel.getCasillaActual();

            String info = "Jugador actual: " + jugador + "\n"
                        + "Casilla: " + casillaActual.getNombre() + "\n"
                        + "Num. propiedades: " + numPropiedades + "\n";

            for (int i = 0; i < numPropiedades; i++)
                info += i + ") " + jugador.getPropiedades().get(i).getNombre() + "\n";	

            info += "\n";
	}
}
